src/System.Management.Automation/engine/remoting/client/remoterunspace.cs

```
@@ -2005,27 +2005,210 @@ public override void StopProcessCommand()
+                        var lineBpParams = new Dictionary<string, object>
+                        {
+                            { "Script", lineBp.Script },
+                            { "Line", lineBp.Line },
+                        };
+
+                        if (lineBp.Column != 0)
+                        {
+                            lineBpParams.Add("Column", lineBp.Column);
+                        }
+
+                        if (lineBp.Action != null)
+                        {
+                            lineBpParams.Add("Action", lineBp.Action);
+                        }
+
+                        InvokeVirtualBreakpointFunction<LineBreakpoint>(DebuggerUtils.SetPSBreakpointFunctionName, lineBpParams);
+                        break;
+
+                    case VariableBreakpoint varBp:
+                        var varBpParams = new Dictionary<string, object>
+                        {
+                            { "Variable", varBp.Variable },
+                            { "Mode", varBp.AccessMode },
+                        };
+
+                        if (varBp.Action != null)
+                        {
+                            varBpParams.Add("Action", varBp.Action);
+                        }
+
+                        if (varBp.Script != null)
+                        {
+                            varBpParams.Add("Script", varBp.Script);
+                        }
+
+                        InvokeVirtualBreakpointFunction<VariableBreakpoint>(DebuggerUtils.SetPSBreakpointFunctionName, varBpParams);
+                        break;
+
+                    default:
+                        // Unreachable default case
+                        Dbg.Assert(false, "Unexpected breakpoint type.");
+                        break;
+                }
+            }
+        }
+
+        public override CommandBreakpoint SetCommandBreakpoint(string command, ScriptBlock action = null, string path = null)
```

You will need to create a capability object that defines what the remote endpoint supports, and if new methods are not supported throw an appropriate error.  We currently use the endpoint PowerShell version value, and I think we can continue doing so ... I don't think we need to pass a separate protocol version data item.

---

Just to make sure I understand, do you mean something like a `RemoteDebugSessionCapability` class, similar to the `RemoteSessionCapability` class? Or is there a better example I should follow for this? A little more detail on what you need would be helpful here.

---

I have been thinking about this.  Currently we use PowerShell version as returned from the endpoint, ServerVersion, to determine what remote debugging features are available on the target and I feel we can continue doing this ... we don't need to have and maintain a separate debugging protocol version.

I do think we should have a `RemoteDebuggingCapapability` class that indicate what remote debugging capabilities each PowerShell version supports.  This can be created on the client remote debugger class based on the ServerVersion returned from the target computer.

---

This is now added. I'm pretty happy with how this centralized MS-PSRDP checks, and how it identifies command names that must be protected across versions.

Note that the existing code does not throw if the remote runspace does not support the `__Set-PSDebuggerStepMode` or `__Set-PSUnhandledBreakpointMode` functions that were added in PSv5. It just returns instead, doing nothing. If we want those to throw instead, using the same logic I use to throw exceptions in the breakpoint functions, I can do that. I wasn't sure though, because unlike the breakpoint functions, those two seem primarily for internal use.

Please let me know what you think about all of this.

---

Yeah, these two APIs just return because an exception can cause a crash or hang during remote debug message processing.  These new breakpoint APIs are meant to be called by clients and so I think they should throw 'unsupported' exceptions.  If we run into problems, we can re-evaluate.